containers/src/Data/IntSet/Internal.hs

```diff
@@ -1477,8 +1478,90 @@ equal _   _   = False
 --------------------------------------------------------------------}
 
 instance Ord IntSet where
-    compare s1 s2 = compare (toAscList s1) (toAscList s2)
-    -- tentative implementation. See if more efficient exists.
+  compare = compareIntSets
+
+compareIntSets :: IntSet -> IntSet -> Ordering
+compareIntSets s1 s2 = case (splitSign s1, splitSign s2) of
+  ((l1, r1), (l2, r2)) -> case go l1 l2 of
+    Less -> LT
+    Prefix' -> if null r1 then LT else GT
+    Equals -> case go r1 r2 of
+      Less -> LT
+      Prefix' -> LT
+      Equals -> EQ
+      FlipPrefix -> GT
+      Greater -> GT
+    FlipPrefix -> if null r2 then GT else LT
+    Greater -> GT
+  where
+    go t1@(Bin p1 l1 r1) t2@(Bin p2 l2 r2) = case treeTreeBranch p1 p2 of
+      ABL -> case go l1 t2 of
+        Prefix' -> Greater
+        Equals -> FlipPrefix
+        o -> o
+      ABR -> Less
+      BAL -> case go t1 l2 of
+        Equals -> Prefix'
+        FlipPrefix -> Less
+        o -> o
+      BAR -> Greater
+      EQL -> case go l1 l2 of
+        Prefix' -> Greater
+        Equals -> go r1 r2
+        FlipPrefix -> Less
+        o -> o
+      NOM -> if unPrefix p1 < unPrefix p2 then Less else Greater
+    go (Bin _ l1 _) (Tip k2 bm2) = case leftmostTipSure l1 of
+      Tip' k1 bm1 -> case orderTips k1 bm1 k2 bm2 of
+        Prefix' -> Greater
+        Equals -> FlipPrefix
+        o -> o
+    go (Tip k1 bm1) (Bin _ l2 _) = case leftmostTipSure l2 of
+      Tip' k2 bm2 -> case orderTips k1 bm1 k2 bm2 of
+        Equals -> Prefix'
+        FlipPrefix -> Less
+        o -> o
+    go (Tip k1 bm1) (Tip k2 bm2) = orderTips k1 bm1 k2 bm2
+    go Nil Nil = Equals
+    go Nil _ = Prefix'
+    go _ Nil = FlipPrefix
+
+-- This type allows GHC to return unboxed ints from leftmostTipSure, as
+-- $wleftmostTipSure :: IntSet -> (# Int#, Word# #)
+-- On a modern enough GHC (>=9.4) this is unnecessary, we could use StrictPair
+-- instead and get the same Core.
+data Tip' = Tip' {-# UNPACK #-} !Int {-# UNPACK #-} !BitMap
+
+leftmostTipSure :: IntSet -> Tip'
+leftmostTipSure (Bin _ l _) = leftmostTipSure l
+leftmostTipSure (Tip k bm) = Tip' k bm
+leftmostTipSure Nil = error "leftmostTipSure: Nil"
+
+orderTips :: Int -> BitMap -> Int -> BitMap -> Order
+orderTips k1 bm1 k2 bm2 = case compare k1 k2 of
+  LT -> Less
+  EQ | bm1 == bm2 -> Equals
+     | otherwise ->
+         let diff = bm1 `xor` bm2
+             lowestDiff = diff .&. negate diff
+             highMask = negate lowestDiff
+         in if bm1 .&. lowestDiff == 0
+            then (if bm1 .&. highMask == 0 then Prefix' else Greater)
+            else (if bm2 .&. highMask == 0 then FlipPrefix else Less)
+  GT -> Greater
+{-# INLINE orderTips #-}
```

Comment:
Please document `leftmostTipSure` and `orderTips`. `orderTips could also use some internal comments to explain what's going on with the bit fiddling.

Reply:
Left a comment in `orderTips`.  
`leftmostTipSure` returns ... the leftmost Tip. I don't know what else to say about that 😄